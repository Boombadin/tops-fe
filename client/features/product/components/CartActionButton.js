import PropTypes from 'prop-types';
import React, { useCallback, useEffect, useState } from 'react';

import { useCartContext, useReduxContext } from '@client/contexts';
import AddProductButton from '@client/features/product/components/AddProductButton';
import AdjustProductQtyBox from '@client/features/product/components/AdjustProductQtyBox';
import { usePrevious } from '@client/hooks';
import {
  notifyMaxQty,
  notifyOutOfStock,
  resetNotifyMaxQty,
  resetNotifyOutOfStock,
} from '@client/reducers/cart';

function CartActionButton({
  addToCartLabel,
  beforeAddProductCallback,
  gtm,
  outOfStockLabel,
  product,
  type,
}) {
  const [currentQty, setCurrentQty] = useState(product.qty);
  const previousQty = usePrevious(currentQty);
  const {
    cartAction: {
      addProduct,
      changeProductQty,
      deleteProduct,
      addDebouncingProduct,
      removeDebouncingProduct,
    },
  } = useCartContext();
  const { reduxAction } = useReduxContext();

  const stockQty = product?.extension_attributes?.stock_item?.qty;
  const maxSaleQty = product?.extension_attributes?.stock_item?.max_sale_qty;

  const onClickAddProductBtn = useCallback(() => {
    setCurrentQty(prevQty => prevQty + 1);
    addProduct(product, {
      beforeAction: beforeAddProductCallback,
    });
  }, [addProduct]);

  const onClickIncreaseQtyBtn = useCallback(() => {
    addDebouncingProduct({ productSku: product.sku });
    setCurrentQty(prevQty => prevQty + 1);
  }, [addDebouncingProduct]);

  const onClickDecreaseQtyBtn = useCallback(() => {
    addDebouncingProduct({ productSku: product.sku });
    setCurrentQty(prevQty => prevQty - 1);
  }, [addDebouncingProduct]);

  const onClickIncreaseQtyDisabledBtn = useCallback(async () => {
    if (currentQty >= stockQty) {
      await reduxAction.dispatch(resetNotifyOutOfStock());
      reduxAction.dispatch(notifyOutOfStock(product.sku));
    } else if (currentQty >= maxSaleQty) {
      await reduxAction.dispatch(resetNotifyMaxQty());
      reduxAction.dispatch(notifyMaxQty(product.sku));
    }
  }, [currentQty, maxSaleQty, stockQty, reduxAction]);

  const onChangeProductQtyCompleted = useCallback(
    newQty => {
      removeDebouncingProduct({ productSku: product.sku });
      changeProductQty({ productSku: product.sku, qty: newQty });
    },
    [changeProductQty, removeDebouncingProduct],
  );

  useEffect(() => {
    if (currentQty === 0 && previousQty > 0) {
      removeDebouncingProduct({ productSku: product.sku });
      deleteProduct({ productSku: product.sku });
    }
  }, [currentQty, previousQty]);

  useEffect(() => {
    setCurrentQty(product.qty);
  }, [product]);

  if (product.qty > 0) {
    return (
      <AdjustProductQtyBox
        defaultQty={product.qty}
        stockQty={stockQty}
        maxSaleQty={maxSaleQty}
        onClickIncreaseQtyBtn={onClickIncreaseQtyBtn}
        onClickDecreaseQtyBtn={onClickDecreaseQtyBtn}
        onClickIncreaseQtyDisabledBtn={onClickIncreaseQtyDisabledBtn}
        onChangeProductQtyCompleted={onChangeProductQtyCompleted}
        gtm={gtm}
        type={type}
      />
    );
  }

  return (
    <AddProductButton
      addToCartLabel={addToCartLabel}
      onClickAddProductBtn={onClickAddProductBtn}
      outOfStockLabel={outOfStockLabel}
      product={product}
      gtm={gtm}
      type={type}
    />
  );
}

AddProductButton.propTypes = {
  beforeAddProductCallback: PropTypes.func,
  gtm: PropTypes.object,
  product: PropTypes.object,
  type: PropTypes.oneOf(['default', 'mini']),
  addToCartLabel: PropTypes.string.isRequired,
  outOfStockLabel: PropTypes.string.isRequired,
};

AddProductButton.defaultProps = {
  beforeAddProductCallback: () => {},
  gtm: {},
  product: {},
  type: 'default',
};

export default React.memo(CartActionButton);
