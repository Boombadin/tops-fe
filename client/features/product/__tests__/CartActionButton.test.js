import { cleanup, fireEvent, render } from '@testing-library/react';
import React from 'react';

import CartActionButton from '@client/features/product/components/CartActionButton';

jest.mock('@client/contexts', () => ({
  useCartContext: jest.fn().mockReturnValue({
    cartAction: {
      addProduct: jest.fn(),
      deleteProduct: jest.fn(),
      addDebouncingProduct: jest.fn(),
      removeDebouncingProduct: jest.fn(),
    },
  }),
  useReduxContext: jest.fn().mockReturnValue({
    reduxAction: {
      dispatch: jest.fn(),
    },
  }),
}));

describe('features/product/components/CartActionButton', () => {
  let props = {};
  beforeEach(() => {
    props = {
      addToCartLabel: 'Add to Cart',
      outOfStockLabel: 'Out of Stock',
      product: {
        qty: 0,
        status: 1,
        visibility: 2,
        extension_attributes: {
          stock_item: { is_in_stock: true, qty: 10 },
        },
      },
    };
  });
  afterEach(cleanup);

  test('it should render AddProductButton incoming if qty <= 0', () => {
    const component = render(<CartActionButton {...props} />);
    component.getByTestId('add-product-button');

    expect(component).toMatchSnapshot();
  });
  test('it should render AdjustProductQtyBox incoming if qty > 0', () => {
    props.product.qty = 1;
    const component = render(<CartActionButton {...props} />);
    component.getByTestId('adjust-product-qty-box');

    expect(component).toMatchSnapshot();
  });
  test('it should render AdjustProductQtyBox instead after click on AddProductButton', async () => {
    const component = render(<CartActionButton {...props} />);
    await component.findByTestId('add-product-button');

    props.product.qty += 1;
    fireEvent.click(component.getByTestId('add-product-button'));
    component.rerender(<CartActionButton {...props} />);
    component.getByTestId('adjust-product-qty-box');
  });
  test('it should render AddProductButton instead when click on decrease qty btn and qty is decreased to 0', async () => {
    props.product.qty = 1;
    const component = render(<CartActionButton {...props} />);
    await component.findByTestId('adjust-product-qty-box');

    fireEvent.click(
      component.getByTestId('adjust-product-qty-box-decrease-btn'),
    );
    const product = { ...product, qty: props.product.qty - 1 };
    component.rerender(<CartActionButton {...props} product={product} />);
    component.getByTestId('add-product-button');
  });
  test('it should render AddProductButton instead if qty in product prop is less than 1', async () => {
    props.product.qty = 10;
    const component = render(<CartActionButton {...props} />);
    await component.findByTestId('adjust-product-qty-box');

    const product = { ...product, qty: 0 };
    component.rerender(<CartActionButton {...props} product={product} />);
    component.getByTestId('add-product-button');
  });
});
