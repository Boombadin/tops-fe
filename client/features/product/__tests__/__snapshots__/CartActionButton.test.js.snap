// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`features/product/components/CartActionButton it should render AddProductButton incoming if qty <= 0 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        class="AddProductButton__Button-ag50oj-0 bkWnZn"
        data-testid="add-product-button"
        type="default"
      >
        Add to Cart
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      class="AddProductButton__Button-ag50oj-0 bkWnZn"
      data-testid="add-product-button"
      type="default"
    >
      Add to Cart
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`features/product/components/CartActionButton it should render AdjustProductQtyBox incoming if qty > 0 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="AdjustProductQtyBox__IncrementInputBoxStyled-sc-15n8obw-0 cACVcL IncrementInputBox__Wrapper-sc-1o57gb-0 XRdDS"
        data-testid="adjust-product-qty-box"
        type="default"
      >
        <div
          class="IncrementInputBox__Element-sc-1o57gb-1 hKTCrY"
        >
          <div
            class="IncrementInputBox__IncButton-sc-1o57gb-4 bLReEN"
          >
            <button
              class="AdjustProductQtyBox__ButtonStyled-sc-15n8obw-1 cBQqxu"
              data-testid="adjust-product-qty-box-decrease-btn"
              type="default"
            >
              -
            </button>
          </div>
        </div>
        <div
          class="IncrementInputBox__Element-sc-1o57gb-1 hKTCrY"
        >
          <div
            class="IncrementInputBox__InputWrapper-sc-1o57gb-2 ieBGjh"
          >
            <input
              class="inputBox IncrementInputBox__InputBox-sc-1o57gb-3 bLAQDR"
              id="inputBox"
              readonly=""
              type="number"
              value="1"
            />
          </div>
        </div>
        <div
          class="IncrementInputBox__Element-sc-1o57gb-1 hKTCrY"
        >
          <div
            class="IncrementInputBox__IncButton-sc-1o57gb-4 bLReEN"
          >
            <button
              class="AdjustProductQtyBox__ButtonStyled-sc-15n8obw-1 cBQqxu"
              data-testid="adjust-product-qty-box-increase-btn"
              type="default"
            >
              +
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="AdjustProductQtyBox__IncrementInputBoxStyled-sc-15n8obw-0 cACVcL IncrementInputBox__Wrapper-sc-1o57gb-0 XRdDS"
      data-testid="adjust-product-qty-box"
      type="default"
    >
      <div
        class="IncrementInputBox__Element-sc-1o57gb-1 hKTCrY"
      >
        <div
          class="IncrementInputBox__IncButton-sc-1o57gb-4 bLReEN"
        >
          <button
            class="AdjustProductQtyBox__ButtonStyled-sc-15n8obw-1 cBQqxu"
            data-testid="adjust-product-qty-box-decrease-btn"
            type="default"
          >
            -
          </button>
        </div>
      </div>
      <div
        class="IncrementInputBox__Element-sc-1o57gb-1 hKTCrY"
      >
        <div
          class="IncrementInputBox__InputWrapper-sc-1o57gb-2 ieBGjh"
        >
          <input
            class="inputBox IncrementInputBox__InputBox-sc-1o57gb-3 bLAQDR"
            id="inputBox"
            readonly=""
            type="number"
            value="1"
          />
        </div>
      </div>
      <div
        class="IncrementInputBox__Element-sc-1o57gb-1 hKTCrY"
      >
        <div
          class="IncrementInputBox__IncButton-sc-1o57gb-4 bLReEN"
        >
          <button
            class="AdjustProductQtyBox__ButtonStyled-sc-15n8obw-1 cBQqxu"
            data-testid="adjust-product-qty-box-increase-btn"
            type="default"
          >
            +
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
